### PR TITLE
add high disk watermark node checks

### PR DESCRIPTION
**Pull request by kovrus**

### CODE REVIEW

File: blackbox/docs/sql/system.txt [hunk 1]
```diff
@@ -665,6 +666,15 @@ Starting with Crate 0.55.0 using Java 7 runtime environment is
 deprecated. Please update to Java 8 (>= update 20). See also `here
 <https://crate.io/docs/reference/en/latest/installation.html#java-jvm-version>`_.
 
+Routing allocation disk watermark high
+......................................
+
+The check for the :ref:`cluster.routing.allocation.disk.watermark.high
+<cluster.routing.allocation.disk.watermark.high>` setting verifies that
+the high watermark is exceeded on the current node. The usage of each disk
```
Comment (seut) [inline blackbox/docs/sql/system.txt]: shouldn't it read: "verifies the high watermark is not exceeded..." ?

File: sql/src/main/java/io/crate/operation/reference/sys/check/node/DiskWatermarkNodesSysCheck.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        } catch (IOException e) {
+            LOGGER.error("Unable to determine the node disk usage while validating high/low disk watermark check: ", e);
+            return false;
+        }
+    }
+
+    protected abstract boolean validate(FsInfo fsInfo, double diskWatermarkPercents, long diskWatermarkBytes);
+
+    private double watermarkSettingInPercents() {
+        return thresholdPercentageFromWatermark(watermarkSetting.extract(settings));
```
Comment (seut) [inline sql/src/main/java/io/crate/operation/reference/sys/check/node/DiskWatermarkNodesSysCheck.java]: afaik there is no other usage of the called method, so why wrap it in another method?

File: sql/src/main/java/io/crate/operation/reference/sys/check/node/DiskWatermarkNodesSysCheck.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    private double thresholdPercentageFromWatermark(String watermark) {
+        try {
+            return RatioValue.parseRatioValue(watermark).getAsPercent();
+        } catch (ElasticsearchParseException ex) {
+            return 100.0;
+        }
+    }
+
+    private long watermarkSettingInBytes() {
+        return thresholdBytesFromWatermark(watermarkSetting.extract(settings));
```
Comment (seut) [inline sql/src/main/java/io/crate/operation/reference/sys/check/node/DiskWatermarkNodesSysCheck.java]: same here

File: sql/src/test/java/io/crate/operation/reference/sys/check/node/SysNodeChecksTest.java [hunk 4]
```diff
@@ -185,4 +190,63 @@ public void testJvmVersion() {
...
+        assertThat(highDiskWatermarkNodesSysCheck.id(), is(5));
+        assertThat(highDiskWatermarkNodesSysCheck.nodeId().utf8ToString(), is("noop_id"));
+        assertThat(highDiskWatermarkNodesSysCheck.severity(), is(SysCheck.Severity.HIGH));
+
+        // disk.watermark.high: 140b
+        when(fsInfo.iterator()).thenReturn(paths.iterator());
+        assertThat(highDiskWatermarkNodesSysCheck.validate(fsInfo, 100.0, 140), is(true));
+
+        // disk.watermark.high: 160b
+        when(fsInfo.iterator()).thenReturn(paths.iterator());
```
Comment (seut) [inline sql/src/test/java/io/crate/operation/reference/sys/check/node/SysNodeChecksTest.java]: why is this mock binding done multiple times?

Comment (kovrus) [inline sql/src/test/java/io/crate/operation/reference/sys/check/node/SysNodeChecksTest.java]: we can use iterator only once

File: sql/src/test/java/io/crate/operation/reference/sys/check/node/SysNodeChecksTest.java [hunk 5]
```diff
@@ -185,4 +190,63 @@ public void testJvmVersion() {
...
+        assertThat(highDiskWatermarkNodesSysCheck.validate(fsInfo, 100.0, 160), is(false));
+
+        // disk.watermark.high: 180b
+        when(fsInfo.iterator()).thenReturn(paths.iterator());
+        assertThat(highDiskWatermarkNodesSysCheck.validate(fsInfo, 100.0, 180), is(false));
```
Comment (seut) [inline sql/src/test/java/io/crate/operation/reference/sys/check/node/SysNodeChecksTest.java]: is this assertion needed? I don't see any difference to the above one

File: sql/src/test/java/io/crate/operation/reference/sys/check/node/SysNodeChecksTest.java [hunk 6]
```diff
@@ -185,4 +190,63 @@ public void testJvmVersion() {
...
+        assertThat(highDiskWatermarkNodesSysCheck.id(), is(5));
+        assertThat(highDiskWatermarkNodesSysCheck.nodeId().utf8ToString(), is("noop_id"));
+        assertThat(highDiskWatermarkNodesSysCheck.severity(), is(SysCheck.Severity.HIGH));
+
+        // disk.watermark.high: 75%
+        when(fsInfo.iterator()).thenReturn(paths.iterator());
+        assertThat(highDiskWatermarkNodesSysCheck.validate(fsInfo, 75.0, 0), is(true));
+
+        // disk.watermark.high: 55%
+        when(fsInfo.iterator()).thenReturn(paths.iterator());
```
Comment (seut) [inline sql/src/test/java/io/crate/operation/reference/sys/check/node/SysNodeChecksTest.java]: again not clear to me why mock binding multiple times is needed ;)

File: sql/src/test/java/io/crate/operation/reference/sys/check/node/SysNodeChecksTest.java [hunk 7]
```diff
@@ -185,4 +190,63 @@ public void testJvmVersion() {
...
+        assertThat(highDiskWatermarkNodesSysCheck.validate(fsInfo, 55.0, 0), is(false));
+
+        // disk.watermark.high: 45%
+        when(fsInfo.iterator()).thenReturn(paths.iterator());
```
Comment (seut) [inline sql/src/test/java/io/crate/operation/reference/sys/check/node/SysNodeChecksTest.java]: and here also, I don't see any need for this additional assertion

Comment (kovrus) [inline sql/src/test/java/io/crate/operation/reference/sys/check/node/SysNodeChecksTest.java]: removing them, agree

File: sql/src/test/java/io/crate/operation/reference/sys/check/node/SysNodeChecksTest.java (hunk 4 above)
Comment (seut) [inline sql/src/test/java/io/crate/operation/reference/sys/check/node/SysNodeChecksTest.java]: ah sure ;)

**Comment by seut** [comment]
lgtm

**Pull request merged**